File: python/muse/data_structures/array_queue.py
class ArrayQueue[T]:
    DEFAULT_CAPACITY = 64

    def __init__(self, physicalSize: int = 0):
        self.data: list[T | None] = []
        self.front = 0
        self.logical_size = 0
        self.physical_size = self.DEFAULT_CAPACITY

        if physicalSize > 1:
            self.physical_size = physicalSize
        self.data = [None] * self.physical_size

    def size(self) -> int:
        return self.logical_size

    def is_empty(self) -> bool:
        return self.logical_size == 0

    def peek(self) -> T:
        if self.logical_size == 0:
            raise RuntimeError("[PANIC - NoSuchElement]")

        return self.data[self.front]

    def enqueue(self, element: T) -> None:
        if self.logical_size == self.physical_size:
            newCapacity = self.DEFAULT_CAPACITY

            if self.physical_size >= self.DEFAULT_CAPACITY:
                newCapacity = self.physical_size + (self.physical_size >> 1)

            temp: list[T | None] = [None] * newCapacity

            cursor = self.front

            for i in range(self.logical_size):
                if cursor == self.physical_size:
                    cursor = 0
                temp[i] = self.data[cursor]
                cursor += 1

            self.data = temp
            self.front = 0
            self.physical_size = newCapacity

        self.data[(self.front + self.logical_size) % self.physical_size] = element

        self.logical_size += 1

    def dequeue(self) -> None:
        if self.logical_size == 0:
            raise RuntimeError("[PANIC - NoSuchElement]")

        self.data[self.front] = None

        self.front = (self.front + 1) % self.physical_size

        self.logical_size -= 1

    def capacity(self) -> int:
        return self.physical_size

    def shrink(self) -> None:
        temp: list[T | None] = [None] * self.logical_size

        cursor = self.front

        for i in range(self.logical_size):
            if cursor == self.physical_size:
                cursor = 0
            temp[i] = self.data[cursor]
            cursor += 1

        self.data = temp
        self.front = 0
        self.physical_size = self.logical_size

File: rust/muse/src/data_structures/array_queue.rs
use super::{Queue, Resizable};

#[derive(Debug)]
pub struct ArrayQueue<T> {
    data: Vec<Option<T>>,
    front: usize,
    logical_size: usize,
    physical_size: usize,
}

impl<T> ArrayQueue<T> {
    const DEFAULT_CAPACITY: usize = 64;

    pub fn new() -> Self {
        Self::new_with_capacity(0)
    }

    pub fn new_with_capacity(physical_size: usize) -> Self {
        let mut queue = Self {
            data: Default::default(),
            front: 0,
            logical_size: 0,
            physical_size: Self::DEFAULT_CAPACITY,
        };

        if physical_size > 1 {
            queue.physical_size = physical_size;
        }
        queue.data = Vec::with_capacity(queue.physical_size);
        (0..queue.physical_size).for_each(|_| {
            queue.data.push(None);
        });

        queue
    }
}

impl<T> Default for ArrayQueue<T> {
    fn default() -> Self {
        Self::new()
    }
}

impl<T> Queue<T> for ArrayQueue<T> {
    fn size(&self) -> usize {
        self.logical_size
    }

    fn is_empty(&self) -> bool {
        self.logical_size == 0
    }

    fn peek(&self) -> &T {
        if self.logical_size == 0 {
            panic!("[PANIC - NoSuchElement]");
        }

        self.data[self.front]
            .as_ref()
            .expect("element should exist")
    }

    fn enqueue(&mut self, element: T) {
        if self.logical_size == self.physical_size {
            let mut new_capacity = Self::DEFAULT_CAPACITY;

            if self.physical_size >= Self::DEFAULT_CAPACITY {
                new_capacity = self.physical_size + (self.physical_size >> 1)
            }

            let mut temp = Vec::with_capacity(new_capacity);

            let mut cursor = self.front;

            (0..self.logical_size).for_each(|_| {
                if cursor == self.physical_size {
                    cursor = 0;
                }
                temp.push(self.data[cursor].take());
                cursor += 1;
            });
            (self.logical_size..new_capacity).for_each(|_| {
                temp.push(None);
            });

            self.data = temp;
            self.front = 0;
            self.physical_size = new_capacity;
        }

        self.data[(self.front + self.logical_size) % self.physical_size] = Some(element);

        self.logical_size += 1;
    }

    fn dequeue(&mut self) {
        if self.logical_size == 0 {
            panic!("[PANIC - NoSuchElement]");
        }

        self.data[self.front].take();

        self.front = (self.front + 1) % self.physical_size;

        self.logical_size -= 1;
    }
}

impl<T> Resizable for ArrayQueue<T> {
    fn capacity(&self) -> usize {
        self.physical_size
    }

    fn shrink(&mut self) {
        let mut temp = Vec::with_capacity(self.logical_size);

        let mut cursor = self.front;

        (0..self.logical_size).for_each(|_| {
            if cursor == self.physical_size {
                cursor = 0;
            }
            temp.push(self.data[cursor].take());
            cursor += 1;
        });

        self.data = temp;
        self.front = 0;
        self.physical_size = self.logical_size;
    }
}

#[cfg(test)]
pub(crate) mod tests;
